vllm/core/scheduler.py

```diff
@@ -32,15 +32,60 @@ class PreemptionMode(enum.Enum):
 @dataclass
 class SchedulingBudget:
     """The available slots for scheduling."""
-    num_batched_tokens: int
-    num_curr_seqs: int
     token_budget: int
     max_num_seqs: int
+    _requeset_ids_num_batched_tokens: Set[int] = field(default_factory=set)
+    _requeset_ids_num_curr_seqs: Set[int] = field(default_factory=set)
+    _num_batched_tokens: int = 0
+    _num_curr_seqs: int = 0
 
     def can_schedule(self, *, num_new_tokens: int, num_new_seqs: int):
+        assert num_new_tokens != 0
+        assert num_new_seqs != 0
         return (self.num_batched_tokens + num_new_tokens <= self.token_budget
                 and self.num_curr_seqs + num_new_seqs <= self.max_num_seqs)
 
+    def remaining_token_budget(self):
+        return self.token_budget - self.num_batched_tokens
+
+    def add_num_batched_tokens(self, seq_group: SequenceGroup,
```

Comment:
1. we should just take req_id if we really need it 
2. do we need it? when would it the case that the same seq group is added multiple time?

Reply:
why would we have duplicated budget add/subtract though? we fully control the scheduler.

Reply:
> we should just take req_id if we really need it

We are using req_id now.

> do we need it? when would it the case that the same seq group is added multiple time?

This happens for normal scheduling path; We add all running num_seqs before scheduling prefill to take into account of it for prefill scheduling (we don't want to have more than max_num_seqs by always prioritizing prefills). 

Originally I assigned them upfront to the budget and schedule_running only subtracted them when preempt happens. But it is more difficult as running contains prefill now (because token size is dynamic). We don't know how many tokens will be available upfront like before.

Reply:
I think other option is that we only assign budget for num_seqs ahead, and schedule_running only update num_batched tokens & only subtract num_seqs, but I thought it is a little bit leaky abstraction. So I went this way

Reply:
Oh if it only happens in the normal scheduling path please explain in comments and add a todo to fully remove it once chunked is the default.

Reply:
```    TODO(sang): Right now, the budget is request_id-aware meaning it can ignore
    budget update from the same request_id. It is because in normal scheduling
    path, we update RUNNING num_seqs ahead of time, meaning it could be
    updated more than once when scheduling RUNNING requests. Since this won't
    happen if we only have chunked prefill scheduling, we can remove this
    feature from the API when chunked prefill is enabled by default.
```

comment added